---
title: "Atlas Integration"
author: "James Ashmore"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Setup

Set chunk options:

```{r knitr}
knitr::opts_chunk$set(
  autodep    = TRUE,
  cache      = FALSE,
  dev        = "png",
  error      = FALSE,
  message    = FALSE,
  warning    = FALSE
)
```

Load Bioconductor packages:

```{r}
library(batchelor)
library(EnsDb.Mmusculus.v79)
library(MouseGastrulationData)
library(org.Mm.eg.db)
library(pheatmap)
library(scater)
library(scran)
```

Load CRAN packages:

```{r}
library(msigdbr)
```

## Data

### Query data

Read experiment object:

```{r}
sce <- readRDS("data/09-cell-annotation.rds")
```

Clear metadata slot:

```{r}
metadata(sce) <- list()
```

Make rownames compatible:

```{r}
rownames(sce) <- rowData(sce)$gene_id
```

Clear rowData slot:

```{r}
rowData(sce) <- NULL
```

Replace rowRanges slot:

```{r}
row <- PartitioningByEnd(integer(length(sce)), names = rownames(sce))

row <- relist(GRanges(), row)

rowRanges(sce) <- row
```

Clear colnames slot:

```{r}
colnames(sce) <- NULL
```

Replace colData slot:

```{r}
colData(sce) <- DataFrame(
  cell = paste("cell", seq_len(ncol(sce)), sep = "_"),
  barcode = NA,
  sample = "LPMP_EHF_B",
  pool = NA,
  stage = "E7.25-8",
  sequencing.batch = NA,
  theiler = "TS11c",
  doub.density = sce$doublet_score,
  doublet = sce$doublet_cluster,
  cluster = sce$cluster,
  cluster.sub = NA,
  cluster.stage = NA,
  cluster.theiler = NA,
  stripped = NA,
  celltype = sce$celltype,
  colour = EmbryoCelltypeColours[sce$celltype]
)
```

Clear reducedDims slot:

```{r}
reducedDims(sce) <- NULL
```

Clear altExp slot:

```{r}
altExp(sce) <- NULL
```

### Reference data

Subset experiment metadata:

```{r}
dat <- subset(AtlasSampleMetadata, stage %in% c("E7.25", "E7.5", "E7.75", "E8.0"))
```

Read experiment object:

```{r}
ref <- EmbryoAtlasData(type = "processed", samples = dat$sample)
```

Filter problematic and unknown cells:

```{r}
nan <- ref$stripped | ref$doublet | is.na(ref$celltype)

ref <- ref[, !nan]
```

Compute log-normalized expression values:

```{r}
ref <- logNormCounts(ref)
```

Clear metadata slot:

```{r}
metadata(ref) <- list()
```

Make rownames compatible:

```{r}
rownames(ref) <- rowData(ref)$ENSEMBL
```

Clear rowData slot:

```{r}
rowData(ref) <- NULL
```

Clear colnames slot:

```{r}
colnames(ref) <- NULL
```

Clear reducedDims slot:

```{r}
reducedDims(ref) <- NULL
```

Clear altExp slot:

```{r}
altExp(ref) <- NULL
```

Split a SingleCellExperiment based on the column data:

```{r}
splitColData <- function(x, f) {
  
  i <- split(seq_along(f), f)

  v <- vector(mode = "list", length = length(i))
  
  names(v) <- names(i)
  
  for (n in names(i)) { v[[n]] <- x[, i[[n]]] }
  
  return(v)

}
```

Split experiment data by the sample column:

```{r}
ref <- splitColData(ref, ref$sample)
```

## Integration

### Combine objects

Create list of experiment data:

```{r}
all <- c(ref, LPMP_EHF_B = sce)
```

```{r include = FALSE}
rm(ref, sce); gc() # delete objects to free memory
```

Subset to common features:

```{r}
ids <- Reduce(intersect, lapply(all, rownames))

all <- lapply(all, "[", i = ids)
```

### Variance modelling

Model the per-gene variance:

```{r}
dec <- lapply(all, modelGeneVar)
```

Combine variance decompositions:

```{r}
dec <- do.call(combineVar, dec)
```

### Gene blacklist

Append Xist gene to blacklist:

```{r}
dec$blacklist <- FALSE

idx <- which(rownames(dec) == "ENSMUSG00000086503")

dec$blacklist[idx] <- TRUE
```

Append Y chromosome genes to blacklist:

```{r}
key <- keys(EnsDb.Mmusculus.v79, keytype = "GENEID", filter = SeqNameFilter("Y"))

idx <- which(rownames(dec) %in% key)

dec$blacklist[idx] <- TRUE
```

Append MT chromosome genes to blacklist:

```{r}
key <- keys(EnsDb.Mmusculus.v79, keytype = "GENEID", filter = SeqNameFilter("MT"))

idx <- which(rownames(dec) %in% key)

dec$blacklist[idx] <- TRUE
```

Append cell-cycle genes to blacklist:

```{r}
sel <- select(org.Mm.eg.db, keys = "GO:0007049", keytype = "GOALL", column = "ENSEMBL")

idx <- which(rownames(dec) %in% sel$ENSEMBL)

dec$blacklist[idx] <- TRUE
```

Append ribosomal genes to blacklist:

```{r}
msd <- msigdbr(species = "Mus musculus", category = "C2")

msd <- subset(msd, gs_name == "KEGG_RIBOSOME")

sel <- select(org.Mm.eg.db, keys = msd$gene_symbol, keytype = "SYMBOL", column = "ENSEMBL")

idx <- which(rownames(dec) %in% sel$ENSEMBL)

dec$blacklist[idx] <- TRUE
```

### Feature selection

Identify highly variable genes:

```{r}
hvg <- subset(dec, blacklist == FALSE)

hvg <- getTopHVGs(hvg, var.field = "bio", var.threshold = 0)
```

### Batch normalization

Perform scaling normalization:

```{r}
all <- do.call(multiBatchNorm, all)
```

Combine objects from experiments:

```{r}
all <- do.call(cbind, all)
```

### Batch integration

Create sequential merge order:

```{r}
dat <- subset(AtlasSampleMetadata, stage %in% all$stage)

lvl <- c("E6.5", "E6.75", "E7.0", "mixed_gastrulation", "E7.25", "E7.5", "E7.75", "E8.0", "E8.25", "E8.5")

dat$stage <- factor(dat$stage, levels = intersect(lvl, all$stage))

dat <- dat[order(dat$stage, dat$ncells, decreasing = TRUE), ]
  
idx <- c(as.character(dat$sample), "LPMP_EHF_B")
```

Correct experiments using nearest neighbours:

```{r}
set.seed(195538969)

mnn <- correctExperiments(
  all,
  batch = all$sample,
  subset.row = hvg,
  correct.all = TRUE,
  PARAM = FastMnnParam(merge.order = idx)
)
```

```{r include = FALSE}
rm(all); gc() # delete objects to free memory
```

## Diagnostics

Check proportion of lost variance:

```{r}
var <- metadata(mnn)$merge.info$lost.var

col <- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = "Reds"))(100)

brk <- seq(0, max(var), length.out = 101)

pheatmap(var, color = col, display_numbers = TRUE, number_format = "%.2f", cluster_cols = FALSE, cluster_rows = FALSE)
```

Cluster on the corrected PCA to obtain a partitioning of the cells:

```{r}
snn <- buildSNNGraph(mnn, type = "jaccard", use.dimred = "corrected")

com <- igraph::cluster_louvain(snn)

mnn$cluster <- factor(com$membership)
```

Tabulate number of cells from multiple batches within each cluster:

```{r}
tab <- table(Cluster = mnn$cluster, Batch = mnn$batch)

col <- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = "Blues"))(100)

brk <- seq(0, max(tab), length.out = 101)

pheatmap(tab, color = col, display_numbers = TRUE, number_format = "%d", cluster_cols = FALSE, cluster_rows = FALSE)
```

Perform UMAP on the corrected PCA matrix:

```{r}
set.seed(42)

mnn <- runUMAP(mnn, dimred = "corrected", n_neighbors = 20, min_dist = 0.7)
```

Visualize the corrected PCA using a UMAP plot:

```{r fig.width = 9, fig.height = 9}
dim <- reducedDim(mnn, "UMAP")

dim <- as.data.frame(dim)

colnames(dim) <- c("x", "y")

col <- colData(mnn)

col <- as.data.frame(col)

dat <- cbind(dim, col)

dim <- dim[sample(nrow(dim), nrow(dim) * 0.1), ]

dat$batch <- factor(dat$batch, levels = idx)

ggplot(dat, aes(x, y)) + 
  geom_point(data = dim, colour = "gainsboro", size = 0.1) + 
  geom_point(aes(colour = celltype), size = 0.1, show.legend = FALSE) + 
  scale_colour_manual(values = EmbryoCelltypeColours) + 
  facet_wrap(~ batch) + 
  theme_bw() + 
  theme(aspect.ratio = 1, axis.title = element_blank())
```

## Post-processing

Make feature names unique:

```{r}
rowData(mnn) <- DataFrame(
  gene_id = rownames(mnn),
  gene_name = mapIds(EnsDb.Mmusculus.v79, keys = rownames(mnn), column = "GENENAME", keytype = "GENEID"),
  rotation = I(rowData(mnn)$rotation)
)
```

Store combined variance model:

```{r}
rownames(dec) <- rownames(mnn)

metadata(mnn)$combineVar <- dec
```

Store highly variable genes:

```{r}
metadata(mnn)$getTopHVGs <- hvg
```

## Summary

### Output

Save integrated experiment object to disk:

```{r}
saveRDS(mnn, file = "data/10-atlas-integration.rds")
```

### Session

Print session information:

```{r}
sessionInfo()
```
