---
title: "iTALK"
author: "James Ashmore"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Setup

Set chunk options:

```{r}
knitr::opts_chunk$set(
  autodep    = TRUE,
  cache      = TRUE,
  cache.path = "cache/italk-analysis.Rmd/",
  dev        = "png",
  error      = FALSE,
  message    = FALSE,
  warning    = FALSE
)
```

## Packages

Load Bioconductor packages:

```{r}
library(scater)
library(scran)
library(scuttle)
```

Load CRAN packages:

```{r}
library(ggforce)
library(ggplot2)
```

Load Github packages:

```{r}
library(iTALK)
```

## Data

Read experiment object:

```{r}
sce <- readRDS("output/secondary-analysis.rds")
```

Read human to mouse annotation table:

```{r}
ann <- read.csv("data/hsapiens_mmusculus.csv")
```

Subset annotation table to orthologous genes:

```{r}
ann <- subset(ann, mmusculus_homology_type == "ortholog_one2one")
```

Identify genes present in experiment object and annotation table:

```{r}
ids <- intersect(rowData(sce)$gene_id, ann$mmusculus_gene_id)
```

Subset experiment object to shared genes:

```{r}
sce <- sce[match(ids, rowData(sce)$gene_id), ]
```

Subset annotation table to shared genes:

```{r}
ann <- ann[match(ids, ann$mmusculus_gene_id), ]
```

Replace row names with human gene identifier:

```{r}
rownames(sce) <- ann$hsapiens_gene_name
```

## Results

Define colour palette:

```{r}
col <- c(
  "1" = "#4E79A7",
  "2" = "#F28E2B",
  "3" = "#E15759",
  "4" = "#76B7B2"
)
```

### Mean count {.tabset}

Parse the data to get Top 50% expressed genes:

```{r}
mat <- logcounts(sce)

mat <- t(mat)

dat <- as.data.frame(mat)

dat <- split(dat, sce$label)

dat <- lapply(dat, colMeans)

dat <- lapply(dat, function(x) {

  d <- data.frame("gene" = names(x), "exprs" = unname(x))
  
  d <- subset(d, exprs > quantile(exprs, prob = 1 - 50/100))

})

dat <- dplyr::bind_rows(dat, .id = "cell_type")

dat <- dat[, c("gene", "exprs", "cell_type")]
```

Find ligand-receptor pairs:

```{r}
res <- lapply(
  X = c("cytokine", "checkpoint", "growth factor", "other"),
  FUN = function(x) {
    
    df <- FindLR(data_1 = dat, datatype = "mean count", comm_type = x)
    
    df$cell_mean_exprs <- df$cell_from_mean_exprs * df$cell_to_mean_exprs
    
    df <- df[order(df$cell_mean_exprs, decreasing = TRUE), ]
    
    df$cell_mean_rank <- rank(-df$cell_mean_exprs)
    
    df
  
  }
)

names(res) <- c("cytokine", "checkpoint", "growth factor", "other")

res <- dplyr::bind_rows(res, .id = "comm_type")
```

#### Cytokine

Plot Top 10 ligand-receptor pairs for "cytokine" communication type:

```{r}
df <- subset(res, comm_type == "cytokine" & cell_mean_rank <= 10)

LRPlot(
  data = df, 
  datatype = "mean count", 
  cell_col = col, 
  link.arr.lwd = df$cell_from_mean_exprs, 
  link.arr.width = df$cell_to_mean_exprs
)
```

#### Checkpoint

Plot Top 10 ligand-receptor pairs for "checkpoint" communication type:

```{r}
df <- subset(res, comm_type == "checkpoint" & cell_mean_rank <= 10)

LRPlot(
  data = df, 
  datatype = "mean count", 
  cell_col = col, 
  link.arr.lwd = df$cell_from_mean_exprs, 
  link.arr.width = df$cell_to_mean_exprs
)
```

#### Growth factor

Plot Top 10 ligand-receptor pairs for "growth factor" communication type:

```{r}
df <- subset(res, comm_type == "growth factor" & cell_mean_rank <= 10)

LRPlot(
  data = df, 
  datatype = "mean count", 
  cell_col = col, 
  link.arr.lwd = df$cell_from_mean_exprs, 
  link.arr.width = df$cell_to_mean_exprs
)
```

#### Other

Plot Top 10 ligand-receptor pairs for "other" communication type:

```{r}
df <- subset(res, comm_type == "other" & cell_mean_rank <= 10)

LRPlot(
  data = df, 
  datatype = "mean count", 
  cell_col = col, 
  link.arr.lwd = df$cell_from_mean_exprs, 
  link.arr.width = df$cell_to_mean_exprs
)
```

### DEG {.tabset}






## Summary

### Session

Print session information:

```{r}
sessionInfo()
```
